xla/pjrt/cpu/mpi_collectives.cc

```diff
@@ -0,0 +1,292 @@
+
+  std::memcpy(output_buffers[rank], input_buffers[rank], chunk_bytes);
+
+  for (int i = 1; i < size; i++) {
+    int send_rank = (rank + i) % size;
+    int recv_rank = (rank + size - i) % size;
+    TF_RETURN_IF_ERROR(MpiErrorToAbslStatus(
+        MPI_Sendrecv(input_buffers[send_rank], chunk_bytes, MPI_BYTE, send_rank,
+                     tag, output_buffers[recv_rank], chunk_bytes, MPI_BYTE,
+                     recv_rank, tag, comm, MPI_STATUS_IGNORE)));
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status MpiCollectivesCommunicator::AllGather(const RendezvousKey& key,
+                                                   size_t chunk_bytes,
+                                                   const void* input_buffer,
+                                                   void* output_buffer,
+                                                   absl::Duration timeout) {
+  return MpiErrorToAbslStatus(MPI_Allgather(input_buffer, chunk_bytes, MPI_BYTE,
+                                            output_buffer, chunk_bytes,
+                                            MPI_BYTE, comm));
+}
+
+absl::Status MpiCollectivesCommunicator::ReduceScatter(
+    const RendezvousKey& key, ReductionKind reduction_kind,
+    PrimitiveType element_type, size_t chunk_elems, const void* input_buffer,
+    void* output_buffer, absl::Duration timeout) {
+  int size;
+  MPI_Comm_size(comm, &size);
+  std::vector<int> recvcounts(size, chunk_elems);
+  TF_ASSIGN_OR_RETURN(MPI_Datatype type, PrimitiveTypeToMpiType(element_type));
+  TF_ASSIGN_OR_RETURN(MPI_Op op, ReductionKindToMpiOp(reduction_kind, type));
+  return MpiErrorToAbslStatus(MPI_Reduce_scatter(
+      input_buffer, output_buffer, recvcounts.data(), type, op, comm));
+}
+
+MpiCollectives::MpiCollectives() {
+  int provided;
+  MPI_Init_thread(NULL, NULL, MPI_THREAD_FUNNELED, &provided);
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+  VLOG(1) << "MPI rank=" << mpi_rank << " size=" << mpi_size;
+}
+
+MpiCollectives::~MpiCollectives() {
+  contexts_.clear();
+  MPI_Finalize();
+}
+
+absl::StatusOr<std::shared_ptr<CollectivesCommunicator>>
+MpiCollectives::GetCommunicator(absl::Span<GlobalDeviceId const> global_devices,
+                                int rank) {
+  int flag;
+  MPI_Is_thread_main(&flag);
+  if (!flag) {
+    return absl::UnknownError(
+        absl::StrCat("MPI: Communicator requested from a thread that is not "
+                     "the one MPI was initialized from. Multiple "
+                     "threads/devices per process are not yet supported."));
+  }
+
+  auto& context = contexts_[std::make_tuple(
+      std::vector<GlobalDeviceId>(global_devices.begin(), global_devices.end()),
+      rank)];
+  if (context) {
+    return context;
+  }
+
+  // we assume that there is only one device per mpi rank
+  // and that the mpi rank and global device id are the identical.
+  TF_RET_CHECK(mpi_rank == global_devices[rank].value());
```

Comment:
This is one of the current limitations, thanks for pointing it out.

If you call [GetTfrtCpuClient](https://github.com/openxla/xla/blob/0bfcaa01d977588d67c672f85855823508382f32/xla/pjrt/cpu/cpu_client.cc#L295) with  `options.node_id` equal to the mpi rank  and `options.cpu_device_count=1` then this assumption is verified. Therefore this seemed like the only natural choice to make in absence of further information on the topology.

However, to make it work in general we would need two maps `mpi_rank[node_id]` and `node_id[global_device_id]`.
The former could be determined easily with an mpi allgather if we pass the node id to the constructor of MpiCollectives. 
The latter will require changing the [CollectivesInterface](https://github.com/openxla/xla/blob/c611014e55e4b28eec2dbd2a6373b89b89350d7b/xla/service/cpu/collectives_interface.h#L78) a bit. I would propose to pass it around via the [CpuExecutableRunOptions](https://github.com/openxla/xla/blob/0bfcaa01d977588d67c672f85855823508382f32/xla/service/cpu/cpu_executable_run_options.h#L26), and add it as an additional (optional) argument to GetCommunicator, then pass it [here](https://github.com/openxla/xla/blob/c611014e55e4b28eec2dbd2a6373b89b89350d7b/xla/service/cpu/cpu_runtime.cc#L472).

wdyt?

(I would like to avoid using the key-value store of the distributed client, since I would like to make this work also without grpc)

Reply:
see comment in main thread below https://github.com/openxla/xla/pull/7849#issuecomment-1890399877